browser/ui/views/location_bar/brave_news_location_view.cc

```diff
@@ -10,210 +10,172 @@
+  auto* contents = GetWebContents();
+  BraveNewsTabHelper* tab_helper =
+      contents ? BraveNewsTabHelper::FromWebContents(contents) : nullptr;
+
+  // When the active tab changes, subscribe to notification when
+  // it has found a feed.
+  if (contents && tab_helper) {
+    // Observe BraveNewsTabHelper for feed changes
+    if (!page_feeds_observer_.IsObservingSource(tab_helper)) {
+      page_feeds_observer_.Reset();
+      page_feeds_observer_.Observe(tab_helper);
+    }
+    // Observe WebContentsObserver for WebContentsDestroyed
+    if (web_contents() != contents) {
+      Observe(contents);
+    }
+  } else {
+    // Unobserve WebContentsObserver
+    if (web_contents()) {
+      Observe(nullptr);
+    }
+    // Unobserve BraveNewsTabHelper
+    if (page_feeds_observer_.IsObserving()) {
+      page_feeds_observer_.Reset();
+    }
   }
 
-  // views::LabelButton:
-  std::unique_ptr<views::LabelButtonBorder> CreateDefaultBorder()
-      const override {
-    std::unique_ptr<views::LabelButtonBorder> border =
-        LabelButton::CreateDefaultBorder();
-    border->set_insets(gfx::Insets::VH(3, 0));
-    return border;
+  // Don't show the icon if preferences don't allow
+  if (!tab_helper || !should_show_.GetValue() || !news_enabled_.GetValue() ||
+      !opted_in_.GetValue()) {
+    SetVisible(false);
+    return;
   }
 
-  std::u16string GetTooltipText(const gfx::Point& p) const override {
-    return l10n_util::GetStringUTF16(IDS_BRAVE_NEWS_ACTION_VIEW_TOOLTIP);
-  }
+  // Verify observing BraveNewsTabHelper
+  DCHECK(page_feeds_observer_.IsObservingSource(tab_helper));
+  // Verify observing for WebContentsDestroyed
+  DCHECK(web_contents());
 
-  SkPath GetHighlightPath() const {
-    gfx::Rect rect(GetPreferredSize());
-    const int radii = ChromeLayoutProvider::Get()->GetCornerRadiusMetric(
-        views::Emphasis::kMaximum, rect.size());
-    SkPath path;
-    path.addRoundRect(gfx::RectToSkRect(rect), radii, radii);
-    return path;
-  }
+  // Icon color changes if any feeds are being followed
+  UpdateIconColor(tab_helper->IsSubscribed());
 
-  // TabStripModelObserver:
-  void OnTabStripModelChanged(
-      TabStripModel* tab_strip_model,
-      const TabStripModelChange& change,
-      const TabStripSelectionChange& selection) override {
-    if (selection.active_tab_changed()) {
-      if (selection.old_contents) {
-        BraveNewsTabHelper::FromWebContents(selection.old_contents)
-            ->RemoveObserver(this);
-      }
-
-      if (selection.new_contents) {
-        BraveNewsTabHelper::FromWebContents(selection.new_contents)
-            ->AddObserver(this);
-      }
-    }
+  // Don't show icon if there are no feeds
+  const bool has_feeds = !tab_helper->GetAvailableFeeds().empty();
+  const bool is_visible = has_feeds || IsBubbleShowing();
+  SetVisible(is_visible);
+}
 
-    Update();
-  }
+void BraveNewsLocationView::WebContentsDestroyed() {
+  page_feeds_observer_.Reset();
```

Comment:
optional: Unobserve the web contents here could be done. Though, WebContentsImpl's observer list doesn't check if observer list is empty on desturction so it's okay to leave as is.